### ISSUE TITLE
Make 'indent' rule configurable w.r.t ternary conditions

### ISSUE THREAD

**Issue by jrajav** [eslint/eslint#6606]
**ESLint version 3.0.0, babel-eslint parser**

**Rule:** http://eslint.org/docs/rules/indent

**Problem:** Ternaries can be formatted better when used for multi-line conditionals, 'indent' rule currently does not allow this.

Example of the formatting pattern:

```js
const object = cond
  ? {
      foo: 'bar'
    }
  : {
      baz: 'qux'
    }
```

Notice how the braces line up and are easy to visually scan compared to this, which is what the 'indent' rule currently expects:

```js
const object = cond
  ? {
    foo: 'bar'
  }
  : {
    baz: 'qux'
  }
```

This goes for parenthetical expressions, multi-line function calls, and array literals as well.

**Solution:** I propose a new option for 'indent' called `"multilineTernary"` that defaults to level 1 like the other opinionated 'indent' options, and which enforces the above pattern rather than the below.

**Comment by platinumazure** [comment]
Just to be clear, is the goal to have the ternary operators be allowed a sort of "negative margin" where they can go left of the expected indent point, as long as the tokens immediately following the ternary operator tokens are aligned correctly?

```js
const object = cond
  ? { foo: 'bar' }
//^ this is inside the indent margin and not warned
//  ^ this is the expected indent point (assuming indent: ["error", 4] config)
  : { baz: 'qux' }
```

Have I captured your proposal correctly here or am I missing something?

**Comment by jrajav** [comment]
Personally, I'd be happy with at least being allowed to indent the extra lines (beyond the ones `?` and `:` live on) one extra level. However, one step further would be to *enforce* the indent at that extra level - no "fuzziness", it's just the new expected indent level if you configure `"multilineTernary"`. So line 3 in my original example has a new expected indent level of 3 instead of 2, and will warn (or error, based on rule level) if it's 2, or 4, etc.

**Comment by platinumazure** [comment]
Ah, now I see, you have a two space configuration in your example. My mistake.

I get the feeling this could be pretty complicated to implement. Let's see what the team thinks.

**Comment by gyandeeps** [comment]
I want to hear what other have to say here but for me the above style is very very unique.
We are trying to solve `multi-line` indentations and I believe introducing ternary option right now will contradict how multi-line behaves. 
Again, I want to hear from other on what their thoughts are.

**Comment by ilyavolodin** [comment]
Agree with @gyandeeps Both of the above examples look really strange to my eyes. I would've made them single line objects to make them more readable. I know it's a matter of personal preference, but unfortunately, we can't support every possible style out there.

**Comment by jrajav** [comment]
@ilyavolodin When it's only a single property, sure - this comes up more when there are multiple properties in an object.

It's probably a style that meshes more with `operator-linebreak: 'before'` and a ternary-heavy codebase, both styles that are more common in React projects where expressions are generally more useful than statements - that is, a more point-free functional style.

It would be pretty disappointing if eslint's built-in rule enforced an opinion on this style... Perhaps if this is thought to be difficult to mesh with the existing rules or doesn't merit its own option, it would be more appropriate to allow setting arbitrary ESTree node types that will trigger a new level of indentation on the following line?

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by jrajav** [comment]
If it's a matter of time commitment, I'd be happy to make a pull request. It sounds like those who weighed in so far wouldn't use this multi-line style to begin with, so changing the behavior of lines following infix-operator-begun lines wouldn't affect them anyway. Can we instead have the discussion from the standpoint of assuming that style, and then what indentation should be enforced for that style?

**Comment by Naoto-Ida** [comment]
It's been over a year, but I think this is still valid.
This is frequently used in RN.
Any way we can figure this out?

**Comment by idbartosz** [comment]
@Naoto-Ida 
Dirty fix is disabling rules with inline comments
```js
/* eslint-disable */
const object = cond
  ? {
      foo: 'bar'
    }
  : {
      baz: 'qux'
    } 
/* eslint-enable */
```

I look forward for this indentation syntax, maybe in 2018 :sweat_smile:
